internal/ingest/event.go

```diff
@@ -0,0 +1,59 @@
+package ingest
+
+import (
+	"context"
+	"encoding/json"
+
+	ingestclient "github.com/artefactual-sdps/enduro/internal/api/gen/http/ingest/client"
+	ingestserver "github.com/artefactual-sdps/enduro/internal/api/gen/http/ingest/server"
+	goaingest "github.com/artefactual-sdps/enduro/internal/api/gen/ingest"
+	"github.com/artefactual-sdps/enduro/internal/event"
+)
+
+// EventSerializer handles serialization/deserialization of ingest events.
+type EventSerializer struct{}
+
+var _ event.Serializer[*goaingest.IngestEvent] = (*EventSerializer)(nil)
+
+func (s *EventSerializer) Marshal(event *goaingest.IngestEvent) ([]byte, error) {
+	return json.Marshal(ingestserver.NewMonitorResponseBody(event))
+}
+
+func (s *EventSerializer) Unmarshal(data []byte) (*goaingest.IngestEvent, error) {
+	payload := ingestclient.MonitorResponseBody{}
+	if err := json.Unmarshal(data, &payload); err != nil {
+		return nil, err
+	}
+	if err := ingestclient.ValidateMonitorResponseBody(&payload); err != nil {
+		return nil, err
+	}
+	return ingestclient.NewMonitorIngestEventOK(&payload), nil
+}
+
+// PublishEvent publishes an ingest event with type safety.
+func PublishEvent(ctx context.Context, svc event.Service[*goaingest.IngestEvent], event any) {
```

Comment:
I wouldn't say that it's type-safe because of the runtime type check that could lead to a panic. Here's an example using a compile-time contraint that you may want to consider: https://gist.github.com/sevein/0cd28e149ab9f4c1317531d9c3da0979.

I've never understood why Goa doesn't export an interface that can be used for cases like this. Maybe there's something that could be done in that area? Perhaps at another time.

Reply:
Thanks @sevein, that looks parfait :ok_hand:!